### ISSUE TITLE
Data Processing for All Relations

### ISSUE THREAD

**Issue by ramarasty** [BunsenFeng/BotHeterogeneity#1]
Thank you for this great work. I have processed Twibot-20 in accordance with the process in BotRGCN. However, this data lacks some of the relation heterogeneity aspects described in the paper (interest domain, counts, and clusters). Is it possible to get that preprocessing code?

**Comment by BunsenFeng** [comment]
Thank you for your interest in our work!

The main experiment (Table 1) only use follower and following relationships. For example, if A follows B, then (A, following, B) and (B, follower, A) are consequently two relational edges.

The "heterogeneity" in this way sounds limited, but that's because the only graph-based dataset TwiBot-20 only supports these two relations. However, our relational graph transformers could be easily extended to more diversified relations. This is why in Table 4, we explored other hand-crafted relations based on existing "follow" edges:

"interest domain": There is a domain field in TwiBot-20, with four cases: D = {politics, entertainment, business, and sports}. We treat each (a,b) \in D \times D as a relation.
"count(status/follower)": Following the approach in [SATAR](https://arxiv.org/abs/2106.13089), we partition users into two classes with the Pareto Principle (the 80/20 rule) based on their num_status and num_follower. We treat (top 20%, top 20%), (top 20%, bottom 80%), (bottom 80%, bottom 80%) as three different relations.
"cluster(bool/description)": We use the True-or-False user property values (bool) or the averaged RoBERTa encoding of user description (description) as user attributes and conduct clustering with K-means (K=3). We treat each (a,b) in {1, 2, 3} \times {1, 2, 3} as a relation.

@TamSiuhin correct me if I am wrong, thank you. Please also help him by uploading the necessary codes.

By the way, we are also working on presenting a new dataset with significantly expanded relation types and node types to better examine the heterogeneities. Stay tuned!

**Comment by ramarasty** [comment]
Thank you for the quick response.
I was not aware that only the follower-following relation was used.
I trained the model in TBD.py, and got around 73% accuracy. The only difference being that I changed the `numeric_features` to 6 like in BotRGCN. I also double checked with the hyperparameter table in the paper.
Tests were also done both on cpu and gpu.

Is there something else that I'm not aware of? Would it help if I increase the number of epochs?

**Comment by TamSiuhin** [comment]
Hi ramarasty, thanks for your interest! I guess this problem is caused by the incorrect input features. Different from BotRGCN, we split the edge index into two files named follower_edge.pt and following_edge.pt and we also perform z-score normalization on numerical features in the data preprocessing step. We will release the source code of data preprocessing in recent days to help you reproduce our results.

**Comment by TamSiuhin** [comment]
Hi @ramarasty, I have uploaded the source code of data processing. It is worth mentioning that user_features2.pt and user_features3.pt are generated the same way as BotRGCN, which can be found at [https://github.com/BunsenFeng/BotRGCN](url). Please feel free to comment if you have more questions.

**Comment by ramarasty** [comment]
@BunsenFeng and @TamSiuhin, thank you so much for the help. It seems to be working great now.
